### PR TITLE
LEP-006: Immediate Close

### PR DESCRIPTION
**Pull request by ambrop72**
LEP for libuv issue https://github.com/libuv/libuv/issues/1536

### CODE REVIEW

File: 006-immediate-close.md [hunk 1]
```diff
@@ -0,0 +1,200 @@
...
+
+An alternative approach would to return `UV_INPROGRESS` and start closing asynchronously
+when `uv_close` is called with null callback but the close cannot be immediate. This
+change would be more backward compatible with applications currently calling `uv_close`
+with null callback. However, it has the issue that an asynchronous close might start
+without a callback so the application would need the ability to set it after caling
+`uv_close`.
+
+Note that if immediate close is successful, callbacks of outstanding requests are
+*not* called. Since the application has closed the handle it is clearly not interested
```
Comment (txdv) [inline 006-immediate-close.md]: maybe call them with ECANCELD?

Comment (ambrop72) [reply 006-immediate-close.md]: I have updated the proposal to mention this possibility, but as noted a decision should not be needed right now.

**Comment by ambrop72** [comment]
Can the responsible persons please comment/accept/reject this proposal?

**Comment by ambrop72** [comment]
Closing due to lack of interest.

**Comment by txdv** [comment]
AFAIK some of the collaborators said that it would change the default behavior too much.

**Comment by PengZheng** [comment]
Not being able to close handle immediately makes object constructor containing several libuv handles hard to write.

Use the following code implementing Actor living within a libuv loop for example:
```C
int UvLoopActorInit(UvLoopActor actor, ActorDispatchMethod dispatch,
                    size_t queueSize, uv_loop_t* loop)
{
	int ret;
	int uv_ret;
	AVER(actor != NULL);
	if(NULL == dispatch || 0 == queueSize || NULL== loop)
		return EDA_INVALID_PARAM;
	(void)queueSize;
	(void)loop;
	ActorInit(&actor->actorStruct, dispatch);
	ret = MessageQueueCreate(&actor->queueStruct, queueSize);
	if(ret != EDA_OK)
	{
		EDA_DBG(RT_ERROR, "Failed to create queue of size %zu (%d)\n",
		        queueSize, ret);
		goto queue_error;
	}
	actor->notify = malloc(sizeof(uv_async_t));
	if(NULL == actor->notify)
	{
		EDA_DBG(SYS_ERROR, "Failed to allocate async\n");
		ret = EDA_NO_MEMORY;
		goto uv_notify_error;
	}
	uv_ret = uv_async_init(loop, actor->notify, OnMessage);
	if(uv_ret)
	{
		EDA_DBG(SYS_ERROR, "Failed to init notify signal %d\n", uv_ret);
		free(actor->notify);
		ret =  EDA_LIBCALL;
		goto uv_notify_error;
	}
	actor->notify->data = actor;
	actor->queueChecker = malloc(sizeof(uv_idle_t));
	if(NULL == actor->queueChecker)
	{
		EDA_DBG(SYS_ERROR, "Failed to allocate idle\n");
		ret = EDA_NO_MEMORY;
		goto idle_alloc_error;
	}
	// it cannot fail, check the libuv documentation
	uv_ret = uv_idle_init(loop, actor->queueChecker);
	AVER(uv_ret == 0);
	actor->queueChecker->data = actor;
	actor->qcheckerActive = false;
	MessageQueueEnable(&actor->queueStruct, true);
	SetClassOfPoly(actor, CLASS(UvLoopActor));
	actor->sig = UvLoopActorSig;
	AVERC(UvLoopActor, actor);
	return EDA_OK;
idle_alloc_error:
    uv_close((uv_handle_t *)actor->notify, (uv_close_cb)free);
uv_notify_error:
	MessageQueueDispose(&actor->queueStruct);
queue_error:
	ActorFinish(&actor->actorStruct);
	return ret;
}
```
`uv_async_t` has to be allocated from some object (heap in this example) living longer than `UvLoopActor`.

If there is good pattern for doing this, please enlighten me.

@ambrop72 @txdv @bnoordhuis

**Pull request closed without merging**